File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #using left and right prefix sum
        answer = [1] * len(nums)

        #compute left
        current_sum = 1
        for idx in range(len(nums)):
            answer[idx] *= current_sum
            current_sum *= nums[idx]
        
        #compute right
        current_sum = 1
        for idx in reversed(range(len(nums))):
            answer[idx] *= current_sum
            current_sum *= nums[idx]

        return answer
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int []answer = new int[nums.length];
        Arrays.fill(answer,1);

        int prev = 1;
        for(int idx=0; idx < answer.length;idx++){
            answer[idx] *= prev;
            prev *= nums[idx];
        }

        prev = 1;
        for(int idx=answer.length-1; idx >= 0;idx--){
            answer[idx] *= prev;
            prev *= nums[idx];
        }

        return answer;

    }
}


        